### PR TITLE
SCP-3196: Added a new Game contract (simplified version of GameStateMachine)

### PR DESCRIPTION
**Pull request by koslambrou**
Added a new Game contract (simplified version of GameStateMachine).

* Parameterized the GameStateMachine contract

* Added the Game contract which is compatible with remote wallets by using `yieldUnbalancedTx`.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-use-cases/src/Plutus/Contracts/GameStateMachine.hs [hunk 1]
```diff
@@ -190,54 +211,55 @@ transition State{stateData=oldData, stateValue=oldValue} input = case (oldData,
...
-mkValidator :: Scripts.ValidatorType GameStateMachine
-mkValidator = SM.mkValidator machine
+mkValidator :: GameParam -> Scripts.ValidatorType GameStateMachine
+mkValidator gameParam = SM.mkValidator (machine gameParam)
 
-typedValidator :: Scripts.TypedValidator GameStateMachine
-typedValidator = Scripts.mkTypedValidator @GameStateMachine
-    $$(PlutusTx.compile [|| mkValidator ||])
+typedValidator :: GameParam -> Scripts.TypedValidator GameStateMachine
+typedValidator gp = Scripts.mkTypedValidator @GameStateMachine
```
Comment (sjoerdvisscher) [inline plutus-use-cases/src/Plutus/Contracts/GameStateMachine.hs:224]: You can use `mkTypedValidatorParam` here.

File: plutus-use-cases/src/Plutus/Contracts/GameStateMachine.hs [hunk 2]
```diff
@@ -163,8 +179,12 @@ data GameInput =
     deriving anyclass (ToJSON, FromJSON)
 
 {-# INLINABLE transition #-}
-transition :: State GameState -> GameInput -> Maybe (TxConstraints Void Void, State GameState)
-transition State{stateData=oldData, stateValue=oldValue} input = case (oldData, input) of
+transition
+    :: GameParam
```
Comment (sjoerdvisscher) [inline plutus-use-cases/src/Plutus/Contracts/GameStateMachine.hs:183]: Shouldn't the `GameParam` be used here, or anywhere else? (Maybe I missed it)

Comment (koslambrou) [reply plutus-use-cases/src/Plutus/Contracts/GameStateMachine.hs:183]: It's actually not used specifically anywhere. I only use it to parameterize the contract (i.e. each game will get a specific script address depending on the input value of `GameParam`). I'll look at if it makes sense to add validation for the `GameParam` input.

Comment (sjoerdvisscher) [reply plutus-use-cases/src/Plutus/Contracts/GameStateMachine.hs:183]: Ah that makes sense. Maybe add a bit of documentation that the point is to get different script addresses.

**Review by sjoerdvisscher — APPROVED** [review]

**Pull request merged**